go/vt/sqlparser/ast.go

```diff
@@ -2541,6 +2541,53 @@ type (
 	JSONUnquoteExpr struct {
 		JSONValue Expr
 	}
+
+	// RegexpInstrExpr represents REGEXP_INSTR()
+	// For more information, visit https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-instr
+	RegexpInstrExpr struct {
+		Expr         Expr
+		Pattern      Expr
+		Position     Expr
+		Occurrence   Expr
+		ReturnOption Expr
+		MatchType    Expr
+	}
+
+	// RegexpLikeExpr represents REGEXP_LIKE()
+	// For more information, visit https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-like
+	RegexpLikeExpr struct {
+		Expr      Expr
+		Pattern   Expr
+		MatchType Expr
+	}
+
+	// RegexpReplaceExpr represents REGEXP_REPLACE()
+	// For more information, visit https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-replace
+	RegexpReplaceExpr struct {
+		Expr       Expr
+		Pattern    Expr
+		Repl       Expr
+		Occurrence Expr
+		Position   Expr
+		MatchType  Expr
+	}
+
+	// RegexpSubstrExpr represents REGEXP_SUBSTR()
+	// For more information, visit https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-substr
+	RegexpSubstrExpr struct {
+		Expr       Expr
+		Pattern    Expr
+		Occurrence Expr
+		Position   Expr
+		MatchType  Expr
+	}
+
+	// RLikeExpr represents RLIKE
+	// For more information, visit https://dev.mysql.com/doc/refman/8.0/en/regexp.html#operator_regexp
+	RLikeExpr struct {
```

Comment:
Do we need / want separate structures for this or do we want to put everything into `REGEXP_LIKE` and always output that format when serializing again?

We also always change `CAST` into `CONVERT` as well for example.

Reply:
> Also, I don't quite understand regarding the CAST into CONVERT thing.

That's an example of where slightly different syntax is also stored in the AST in the same structure.

Reply:
I agree with @dbussink, we should normalise into the same AST struct, which I see we have already done now

Reply:
@K-Kumar-01 Oh no! This won't work. I see in the docs https://dev.mysql.com/doc/refman/5.7/en/regexp.html, that mysql 5.7 didn't have the `regexp_like` function at all. So we can't normalise all the inputs to that function call blindly, otherwise things that used to work with mysql 5.7 will start failing 😢

Reply:
There are 2 options here - 
1. Go back to having two different structs, one for `regexp_like` and one for `rlike` and `regexp`
2. Have one struct, but check whether Match_type is empty in `regexp_like` and if it is, print it out as `regexp` and not a function call.

I think both solutions have there own issues, the first one will increase the number of structs we have and maybe a little harder to use later on. The second solution will mean that we print out `regexp` whenever we can, and use `regexp_like` only when we need to. I don't know if this conversion between the two will always be safe or not, in nested function calls, etc. We tried to do something similar with printing of brackets in default clause, and it didn't work out all that well with us finding multiple bugs in it later on.

WDYT @vitessio/query-serving?

Reply:
> Go back to having two different structs, one for regexp_like and one for rlike and regexp

Seems like this is more robust and when we get to the future point of deprecating 5.7 support, we can clean it up then if we want to / if it's a significant burden?

Reply:
@dbussink @GuptaManan100 
The earlier implementation for `regex` didn't have any struct. It was implemented as:
```
| bit_expr REGEXP bit_expr
  {
	$$ = &ComparisonExpr{Left: $1, Operator: RegexpOp, Right: $3}
  }
| bit_expr NOT REGEXP bit_expr
  {
	 $$ = &ComparisonExpr{Left: $1, Operator: NotRegexpOp, Right: $4}
  }
```
So, I was thinking of not creating the struct for Rlike. Instead extend this parsing to support Rlike expressions.
Should I go ahead with this implementation?